File: uri/ad-hock/parking_lot/1246.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int c, n;
    while (cin >> c >> n) {
        vector<int> slots(c, 0);
        unordered_map<int, pair<int, int>> parked;
        int revenue = 0;
        for (int i = 0; i < n; ++i) {
            char type;
            cin >> type;
            if (type == 'C') {
                int plate, length;
                cin >> plate >> length;
                int start = -1;
                for (int pos = 0; pos + length <= c; ++pos) {
                    bool free = true;
                    for (int k = 0; k < length; ++k) {
                        if (slots[pos + k] != 0) {
                            free = false;
                            break;
                        }
                    }
                    if (free) {
                        start = pos;
                        break;
                    }
                }
                if (start != -1) {
                    for (int k = 0; k < length; ++k) {
                        slots[start + k] = plate;
                    }
                    parked[plate] = {start, length};
                    revenue += 10;
                }
            } else if (type == 'S') {
                int plate;
                cin >> plate;
                auto info = parked[plate];
                for (int k = 0; k < info.second; ++k) {
                    slots[info.first + k] = 0;
                }
                parked.erase(plate);
            }
        }
        cout << revenue << '\n';
    }

    return 0;
}

File: uri/ad-hock/parking_lot/1246.rs
use std::collections::HashMap;
use std::io::{self, Read};

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut iter = input.split_whitespace();
    let mut outputs = Vec::new();
    loop {
        let c_value = match iter.next() {
            Some(token) => token,
            None => break,
        };
        let c: usize = c_value.parse().unwrap();
        let n: usize = iter.next().unwrap().parse().unwrap();
        let mut slots = vec![0i32; c];
        let mut parked: HashMap<i32, (usize, usize)> = HashMap::new();
        let mut revenue = 0;
        for _ in 0..n {
            let event = iter.next().unwrap();
            if event == "C" {
                let plate: i32 = iter.next().unwrap().parse().unwrap();
                let length: usize = iter.next().unwrap().parse().unwrap();
                let mut start = None;
                for pos in 0..=c.saturating_sub(length) {
                    let mut free = true;
                    for k in 0..length {
                        if slots[pos + k] != 0 {
                            free = false;
                            break;
                        }
                    }
                    if free {
                        start = Some(pos);
                        break;
                    }
                }
                if let Some(pos) = start {
                    for k in 0..length {
                        slots[pos + k] = plate;
                    }
                    parked.insert(plate, (pos, length));
                    revenue += 10;
                }
            } else {
                let plate: i32 = iter.next().unwrap().parse().unwrap();
                if let Some((pos, length)) = parked.remove(&plate) {
                    for k in 0..length {
                        slots[pos + k] = 0;
                    }
                }
            }
        }
        outputs.push(revenue.to_string());
    }
    println!("{}", outputs.join("\n"));
}
